cmd/util/ledger/migrations/account_key_deduplication_encoder.go

```diff
@@ -0,0 +1,414 @@
+//
+// With deduplicated flag, account status is encoded as:
+// - account status v3 (29 bytes)
+// - length prefixed list of account public key weight and revoked status starting from key index 1
+// - startKeyIndex (4 bytes) + length prefixed list of account public key index mappings to stored key index
+// - startStoredKeyIndex (4 bytes) + length prefixed list of last N stored key digests
+//
+// Without deduplicated flag, account status is encoded as:
+// - account status v3 (29 bytes)
+// - length prefixed list of account public key weight and revoked status starting from key index 1
+// - startStoredKeyIndex (4 bytes) + length prefixed list of last N stored key digests
+func encodeAccountStatusV4WithPublicKeyMetadata(
+	original []byte,
+	weightAndRevokedStatus []accountPublicKeyWeightAndRevokedStatus,
+	startKeyIndexForDigests uint32,
+	keyDigests []uint64,
+	startKeyIndexForMappings uint32,
+	accountPublicKeyMappings []uint32,
+	deduplicated bool,
+) ([]byte, error) {
+
+	// Encode list of account public key weight and revoked status
+	encodedAccountPublicKeyWeightAndRevokedStatus, err := encodeAccountPublicKeyWeightsAndRevokedStatus(weightAndRevokedStatus)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encode list of key digests
+	encodedKeyDigests := encodeDigestList(keyDigests)
+
+	// Encode mappings for deduplicated account status
+	var encodedAccountPublicKeyMapping []byte
+	if deduplicated {
+		encodedAccountPublicKeyMapping, err = encodeAccountPublicKeyMapping(accountPublicKeyMappings)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	newAccountStatusPayloadSize := len(original) +
+		lengthPrefixSize + len(encodedAccountPublicKeyWeightAndRevokedStatus) + // length prefixed account public key weight and revoked status
+		4 + // start stored key index for digests
+		lengthPrefixSize + len(encodedKeyDigests) // length prefixed digests
+
+	if deduplicated {
+		newAccountStatusPayloadSize += 4 + // start key index for mapping
+			lengthPrefixSize + len(encodedAccountPublicKeyMapping) // used to retrieve account public key
+	}
+
+	buf := make([]byte, newAccountStatusPayloadSize)
+	off := 0
+
+	// Append account status v4 version and flag
+	if deduplicated {
+		buf[0] = deduplicatedAccountStatusV4VerionAndFlagByte
+	} else {
+		buf[0] = nondeduplicatedAccountStatusV4VerionAndFlagByte
+	}
+	off++
+
+	// Append original content, except for the flag byte
+	n := copy(buf[off:], original[1:])
+	off += n
+
+	// Append length prefixed encoded revoked status
+	binary.BigEndian.PutUint32(buf[off:], uint32(len(encodedAccountPublicKeyWeightAndRevokedStatus)))
+	off += 4
+
+	n = copy(buf[off:], encodedAccountPublicKeyWeightAndRevokedStatus)
+	off += n
+
+	if deduplicated {
+		// Append start key index for mapping
+		binary.BigEndian.PutUint32(buf[off:], startKeyIndexForMappings)
+		off += 4
+
+		// Append length prefixed account public key mapping
+		binary.BigEndian.PutUint32(buf[off:], uint32(len(encodedAccountPublicKeyMapping)))
```

Comment:
This function contains lots of manual encoding work, I wonder if possible the encoding could be done by a library? So that we could move most logic onto working with higher level structured data, and let the codec library does the encoding and decoding work. It could have better abstraction layer.